File: src/main/java/com/refinedmods/refinedstorage/item/blockitem/BaseBlockItem.java
package com.refinedmods.refinedstorage.item.blockitem;

import com.refinedmods.refinedstorage.block.BaseBlock;
import com.refinedmods.refinedstorage.block.BlockDirection;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.block.state.BlockState;

public class BaseBlockItem extends BlockItem {
    private final BaseBlock block;

    public BaseBlockItem(BaseBlock block, Properties builder) {
        super(block, builder);

        this.block = block;
    }

    @Override
    protected boolean placeBlock(BlockPlaceContext context, BlockState state) {
        boolean result = super.placeBlock(context, state);

        if (result && block.getDirection() != BlockDirection.NONE) {
            context.getLevel().setBlockAndUpdate(context.getClickedPos(), state.setValue(block.getDirection().getProperty(), block.getDirection().getFrom(
                context.getClickedFace(),
                context.getClickedPos(),
                context.getPlayer()
            )));
        }

        return result;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/item/blockitem/BaseBlockItem.kt
package com.refinedmods.refinedstorage.item.blockitem

import com.refinedmods.refinedstorage.block.BaseBlock
import com.refinedmods.refinedstorage.block.BlockDirection
import net.minecraft.block.BlockState
import net.minecraft.item.BlockItem
import net.minecraft.item.ItemPlacementContext

open class BaseBlockItem(private val block: BaseBlock, builder: Settings) : BlockItem(block, builder) {

    override fun place(context: ItemPlacementContext, state: BlockState): Boolean {
        val result: Boolean = super.place(context, state)
        if (result && block.direction !== BlockDirection.NONE) {
            val newState = state.with(
                    block.direction.property,
                    block.direction.getFrom(
                            context.side,
                            context.blockPos,
                            context.player!! // TODO This isn't safe!
                    )
            )

            context.world.setBlockState(context.blockPos, newState)
        }
        return result
    }

}